Replicaset pretty log 1761202

jameinel opened the pull request:
Prettier log messages so that it is more understandable to humans what is going on.

1. Continually sort config.Members (using sort.Stable means it should do a single pass over already sorted data). But this means that when showing the output the "Id" values are kept in sorted order, so one config isn't displayed as 1,2,3, and the other as 1,3,2 making it much harder to visually compare.
2. Tweak fmtConfigForLog so that it doesn't try to squeeze the whole structure onto a single line
3. Include whether on not the Member is considered 'voting'. It was missed in the log output before.
4. I tested using kr/pretty.Diff, but it didn't give very helpful output. If we had a 'unified diff' that we were using, we could probably run it on the textual format, and get a pretty nice diff. But I don't think we have a standard one.

The formatting of new members changed from something like:
```
machine-0: 18:08:29 DEBUG juju.replicaset Set() changing replica set
from {Name: juju, Version: 10, Members: {Member{2 "10.16.17.4:37017" map[juju-machine-id:2]}, Member{1 "10.16.17.189:37017" map[juju-machine-id:0]}, Member{3 "10.16.17.165:37017" map[juju-machine-id:3]}}}
  to {Name: juju, Version: 11, Members: {Member{3 "10.16.17.165:37017" map[juju-machine-id:3]}, Member{2 "10.16.17.4:37017" map[juju-machine-id:2]}, Member{1 "10.16.17.189:37017" map[juju-machine-id:0]}}}
```
To something like:
```
machine-2: 13:11:03 DEBUG juju.replicaset Set() changing replica set
from {
  Name: juju,
  Version: 1,
  Members: {
    {1 "10.16.17.189:37017" juju-machine-id:0 voting},
  },
}
to {
  Name: juju,
  Version: 2,
  Members: {
    {1 "10.16.17.189:37017" juju-machine-id:0 voting},
    {2 "10.16.17.8:37017" juju-machine-id:1 not-voting},
    {3 "10.16.17.224:37017" juju-machine-id:2 not-voting},
  },
}
```
and
```
machine-2: 13:12:03 DEBUG juju.replicaset Set() changing replica set
from {
  Name: juju,
  Version: 2,
  Members: {
    {1 "10.16.17.189:37017" juju-machine-id:0 voting},
    {2 "10.16.17.8:37017" juju-machine-id:1 not-voting},
    {3 "10.16.17.224:37017" juju-machine-id:2 not-voting},
  },
}
to {
  Name: juju,
  Version: 3,
  Members: {
    {1 "10.16.17.189:37017" juju-machine-id:0 voting},
    {2 "10.16.17.8:37017" juju-machine-id:1 voting},
    {3 "10.16.17.224:37017" juju-machine-id:2 voting},
  },
}
```

Also the initialization step changed from
```
[LOG] 0:04.100 INFO juju.replicaset Initiating replicaset with config: replicaset.Config{Name:"juju", Version:1, Members:[]replicaset.Member{replicaset.Member{Id:1, Address:"localhost:38134", Priority:(*float64)(nil), Tags:map[string]string{"foo":"bar"}, Votes:(*int)(nil)}}}
```
to
```
[LOG] 0:04.079 INFO juju.replicaset Initiating replicaset with config: {
  Name: juju,
  Version: 1,
  Members: {
    {1 "localhost:45983" foo:bar voting},
  },
}
```
https://bugs.launchpad.net/juju/+bug/1761202

jameinel committed:
Working on prettier Member formatting.

jameinel committed:
More cleanup of the formatting of Members.

Now we sort them, and hopefully keep them sorted over multiple
calls.

jameinel committed:
pretty.Diff wasn't actually helpful.

When using pretty.Diff, it didn't do a great job of conveying things
like "the new struct has 2 new Members". So we're instead just printing
out the individual lines.

There doesn't seem to be a stdlib 'unified diff' which would be the
other way that we could do some nice formatting. I don't think juju
already has a unidiff dependency, and I don't really want to add one.

jameinel committed:
left justifying the 'to' because it is now multiline

manadart reviewed (approved):
I would drop the right-justify from "to" and just put it under the final brace of "from".

jameinel commented:
$$merge$$

jameinel committed:
Don't mutate the original slice in fmt, but do sort more often.

Sort before we Set() and update the test to understand that we don't
preserve the passed in order, but we Sort() the values.

jameinel commented:
![](https://github.trello.services/images/mini-trello-icon.png) [1761202 juju.replicaset needs better logging](https://trello.com/c/cSJ61c2e/147-1761202-jujureplicaset-needs-better-logging)

jameinel committed:
We no longer mutate the caller's struct.

jameinel commented:
$$merge$$
